### PR TITLE
Fix codegen to ignore httpLabel on non-input shapes per Smithy spec

### PR DESCRIPTION
**Pull request by RanVaknin**
### Context
Per the [Smithy spec](https://smithy.io/2.0/spec/http-bindings.html#httplabel-is-only-used-on-top-level-input), httpLabel on non-input shapes "has no meaning and is simply ignored." However, `AddShapes.findRequestUri()` was throwing a `ModelInvalidException` when it found a shape with a URI bound member that wasn't a direct operation input.

When the shape isn't a direct operation input, it now returns empty instead of throwing (results in codegen ignoring this member). The existing validation for operations with a missing `requestUri` is unchanged.

### Test coverage matrix
The following table shows whether a combination of location / shape has meaning in codegen or is ignored (member gets serialized in PAYLOAD): 

| Location | On Input | On Output | On Error | On Nested |
|---|---|---|---|---|
| URI | pre-existing, honored | added, ignored | added, ignored | added, ignored |
| QUERY_STRING | pre-existing, honored | added, ignored | added, ignore | added, ignored |
| HEADER | pre-existing, honored | added, honored | added, honored | added, ignored |
| HEADERS | same path as HEADER | same path as HEADER | same path as HEADER | added, ignored |
| STATUS_CODE | added, ignored | added, honored | added, ignored | added, ignored |

### CODE REVIEW

File: codegen/src/main/java/software/amazon/awssdk/codegen/AddShapes.java [hunk 1]
```diff
@@ -342,38 +342,54 @@ private boolean isRequiredMember(String memberName, Shape memberShape) {
...
-                        .orElseThrow(() -> {
-                            String detailMsg = "Could not find request URI for input shape for operation: " + operation;
-                            ValidationEntry entry =
-                                new ValidationEntry().withErrorId(ValidationErrorId.REQUEST_URI_NOT_FOUND)
-                                                     .withDetailMessage(detailMsg)
-                                                     .withSeverity(ValidationErrorSeverity.DANGER);
-                            return ModelInvalidException.builder().validationEntries(Collections.singletonList(entry)).build();
-                        });
+        if (!operation.isPresent()) {
+            // Not a direct operation input shape, should be ignored.
```
Comment (joviegas) [inline codegen/src/main/java/software/amazon/awssdk/codegen/AddShapes.java:418]: How does generated code looks like when we have below 
```json
  "NestedOptions": {
      "type": "structure",
      "members": {
        "pageSize": {
          "shape": "String",
          "location": "uri",
          "locationName": "pageSize"
        }
      }
    }
```

when we say
```
httpLabel is only considered when applied to top-level members of an operation's input structure. This trait has no meaning in any other context and is simply ignored.
```

what does ignore mean ? should we internally treat it as 
```
  "NestedOptions": {
      "type": "structure",
      "members": {
        "pageSize": {
          "shape": "String",
          "locationName": "pageSize"
        }
      }
    }
```
and marshall it in the  regular payload. , just as we do for shapes where locationName is not specified

Or completely remove pageSize from NestedOptions structure ?

Can we please have a test case comparing a generated code  ?

Comment (RanVaknin) [reply codegen/src/main/java/software/amazon/awssdk/codegen/AddShapes.java:418]: added a fixture file showing the generated code results in this information getting "ignored" per the smithy spec, and defaulting to serialization via Payload instead.

File: codegen/src/main/java/software/amazon/awssdk/codegen/AddShapes.java [hunk 2]
```diff
@@ -342,38 +367,52 @@ private boolean isRequiredMember(String memberName, Shape memberShape) {
...
+            return Optional.empty();
+        }
+
+        String requestUri = operation.get().getHttp().getRequestUri();
+        if (requestUri == null) {
+            String shapeName = allC2jShapes.entrySet().stream()
+                .filter(e -> e.getValue().equals(parentShape))
+                .map(Map.Entry::getKey)
+                .findFirst()
+                .get();
```
Comment (joviegas) [inline codegen/src/main/java/software/amazon/awssdk/codegen/AddShapes.java]: If parentShape is not in allC2jShapes (edge case: corrupted model), .get() throws NPE, masking the real  issue.

```java
.findFirst()
.orElse("<unknown shape>");

```

File: codegen/src/test/java/software/amazon/awssdk/codegen/AddShapesTest.java [hunk 3]
```diff
@@ -84,7 +84,7 @@ void generateShapeModel_memberRequiredByNestedShape_setsMemberModelAsRequired()
         MemberModel requiredMemberModel = requestShapeModel.findMemberModelByC2jName(queryParamName);
 
         assertThat(requestShapeModel.getRequired()).contains(queryParamName);
-        assertThat(requiredMemberModel.getHttp().getLocation()).isEqualTo(Location.QUERY_STRING);
+        assertThat(requiredMemberModel.getHttp().getLocation()).isNull();
```
Comment (joviegas) [inline codegen/src/test/java/software/amazon/awssdk/codegen/AddShapesTest.java:87]: Rather than modifying this existing assertion, could we keep the original test as-is and add a **new**  operation/input shape that exercises the nested-location-ignored scenario? That way the existing test continues to serve as a regression guard for its original intent, and the new test explicitly documents the Smithy "simply ignored" behavior. Changing a passing assertion to a different expected value makes it harder to tell if we broke something vs. intentionally changed behavior.

File: codegen/src/main/java/software/amazon/awssdk/codegen/AddShapes.java [hunk 4]
```diff
@@ -323,6 +328,26 @@ private ParameterHttpMapping generateParameterHttpMapping(Shape parentShape,
         return mapping;
     }
 
+    private boolean isDirectOperationInputOrOutput(Shape parentShape, Map<String, Shape> allC2jShapes) {
```
Comment (joviegas) [inline codegen/src/main/java/software/amazon/awssdk/codegen/AddShapes.java]: Nit or Good to have : This might be helpful for services with huge models we can try to cache the directOperationShapes in AddShapes
    
private Set<Shape> directOperationShapes;  
```java
private boolean isDirectOperationInputOrOutput(Shape parentShape, Map<String, Shape> allC2jShapes) {
    if (directOperationShapes == null) {
        directOperationShapes = new HashSet<>();
        for (Operation op : builder.getService().getOperations().values()) {
            if (op.getInput() != null) {
                directOperationShapes.add(allC2jShapes.get(op.getInput().getShape()));
            }
            if (op.getOutput() != null) {
                directOperationShapes.add(allC2jShapes.get(op.getOutput().getShape()));
            }
        }
    }
    return directOperationShapes.contains(parentShape);
}

```
Good to have , we can do a quick compilation with ec2 which has approximately 3k shapes and see if it causes any difference

File: codegen/src/test/resources/software/amazon/awssdk/codegen/uri-on-non-input-shape-service.json [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+      }
+    }
+  },
+  "shapes": {
+    "SomeOperationRequest": {
+      "type": "structure",
+      "members": {
+        "thingId": {
+          "shape": "String",
+          "location": "uri",
```
Comment (joviegas) [inline codegen/src/test/resources/software/amazon/awssdk/codegen/uri-on-non-input-shape-service.json:44]: Can we add test cases for the other location types (`httpHeader`, `httpQuery`, `httpPrefixHeaders`)  and also a case where a shape with HTTP bindings is shared between a direct operation input/output and  a non-direct nested context — verifying the location is preserved in the former and ignored in the latter?

File: codegen/src/main/java/software/amazon/awssdk/codegen/AddShapes.java (hunk 4 above)
Comment (RanVaknin) [reply codegen/src/main/java/software/amazon/awssdk/codegen/AddShapes.java]: ec2 took 46seconds to generate, and 40 seconds with this cached approach. not really that noticeable.

File: codegen/src/main/java/software/amazon/awssdk/codegen/AddShapes.java (hunk 2 above)
Comment (RanVaknin) [reply codegen/src/main/java/software/amazon/awssdk/codegen/AddShapes.java]: ended up doing  .orElseThrow instead (this follows existing codegen patterns)

File: codegen/src/main/java/software/amazon/awssdk/codegen/AddShapes.java [hunk 6]
```diff
@@ -311,8 +315,13 @@ private ParameterHttpMapping generateParameterHttpMapping(Shape parentShape,
 
         ParameterHttpMapping mapping = new ParameterHttpMapping();
 
+        // https://smithy.io/2.0/spec/http-bindings.html#httplabel-is-only-used-on-top-level-input
```
Comment (joviegas) [inline codegen/src/main/java/software/amazon/awssdk/codegen/AddShapes.java]: This link is specific to `http-label` but this code handles all the locations types can we say in comment about this and use link https://smithy.io/2.0/spec/http-bindings.html

File: codegen/src/test/java/software/amazon/awssdk/codegen/AddShapesTest.java [hunk 7]
```diff
@@ -84,8 +84,22 @@ void generateShapeModel_memberRequiredByNestedShape_setsMemberModelAsRequired()
         MemberModel requiredMemberModel = requestShapeModel.findMemberModelByC2jName(queryParamName);
 
         assertThat(requestShapeModel.getRequired()).contains(queryParamName);
-        assertThat(requiredMemberModel.getHttp().getLocation()).isEqualTo(Location.QUERY_STRING);
```
Comment (joviegas) [inline codegen/src/test/java/software/amazon/awssdk/codegen/AddShapesTest.java:87]: why are we removing this assertion ?

File: codegen/src/test/resources/software/amazon/awssdk/codegen/poet/model/nestedqueryparameteroperation.java [hunk 8]
```diff
@@ -33,13 +33,13 @@ public final class NestedQueryParameterOperation implements SdkPojo, Serializabl
         .memberName("QueryParamOne")
         .getter(getter(NestedQueryParameterOperation::queryParamOne))
         .setter(setter(Builder::queryParamOne))
-        .traits(LocationTrait.builder().location(MarshallLocation.QUERY_PARAM).locationName("QueryParamOne").build(),
+        .traits(LocationTrait.builder().location(MarshallLocation.PAYLOAD).locationName("QueryParamOne").build(),
```
Comment (joviegas) [inline codegen/src/test/resources/software/amazon/awssdk/codegen/poet/model/nestedqueryparameteroperation.java]: I see here the MarshallLocation of existing test case is getting changed , any reason on why the existing shape is getting changed ?

Comment (RanVaknin) [reply codegen/src/test/resources/software/amazon/awssdk/codegen/poet/model/nestedqueryparameteroperation.java]: Same reason as above, "location":"querystring" on a nested member that is not an input/output/errors should go in the payload.

File: codegen/src/test/java/software/amazon/awssdk/codegen/AddShapesTest.java (hunk 7 above)
Comment (RanVaknin) [reply codegen/src/test/java/software/amazon/awssdk/codegen/AddShapesTest.java:87]: Because it was asserting the buggy behavior. `NestedQueryParameterOperation` is not a direct operation input/output/errors.
I should have added an assert null instead (null would result in the default behavior of it being serialized to the payload instead)

File: codegen/src/main/java/software/amazon/awssdk/codegen/AddShapes.java (hunk 6 above)
Comment (RanVaknin) [reply codegen/src/main/java/software/amazon/awssdk/codegen/AddShapes.java]: fixed

File: codegen/src/test/java/software/amazon/awssdk/codegen/AddShapesTest.java (hunk 7 above)
Comment (joviegas) [reply codegen/src/test/java/software/amazon/awssdk/codegen/AddShapesTest.java:87]: Thanks Ran

**Review by joviegas — APPROVED** [review]

File: codegen/src/main/java/software/amazon/awssdk/codegen/AddShapes.java [hunk 9]
```diff
@@ -323,6 +327,45 @@ private ParameterHttpMapping generateParameterHttpMapping(Shape parentShape,
...
+                      .anyMatch(o -> allC2jShapes.get(o.getInput().getShape()).equals(shape));
+    }
+
+    private boolean isDirectOutputShape(Shape shape, Map<String, Shape> allC2jShapes) {
+        return builder.getService().getOperations().values().stream()
+                      .filter(o -> o.getOutput() != null)
+                      .anyMatch(o -> allC2jShapes.get(o.getOutput().getShape()).equals(shape));
+    }
+
+    private boolean isDirectOperationShape(Shape shape, Map<String, Shape> allC2jShapes) {
```
Comment (joviegas) [inline codegen/src/main/java/software/amazon/awssdk/codegen/AddShapes.java]: Nit: Can we rename this to isDirectInputOutputOrErrorShape

**Review by joviegas — APPROVED** [review]

**Review by joviegas — CHANGES_REQUESTED** [review]
For the changes you discussed offline

File: codegen/src/main/java/software/amazon/awssdk/codegen/AddShapes.java [hunk 10]
```diff
@@ -311,18 +311,67 @@ private ParameterHttpMapping generateParameterHttpMapping(Shape parentShape,
 
         ParameterHttpMapping mapping = new ParameterHttpMapping();
 
+        // Per the Smithy spec, HTTP binding traits are only honored on specific shape types.
```
Comment (alextwoods) [inline codegen/src/main/java/software/amazon/awssdk/codegen/AddShapes.java]: "specific shape types" feels a little vague.  Isn't this just non-top level shapes?

File: codegen/src/main/java/software/amazon/awssdk/codegen/AddShapes.java [hunk 11]
```diff
@@ -311,18 +311,67 @@ private ParameterHttpMapping generateParameterHttpMapping(Shape parentShape,
...
+        // When a trait is ignored, its locationName is also ignored so the member name is used as the wire name.
```
Comment (alextwoods) [inline codegen/src/main/java/software/amazon/awssdk/codegen/AddShapes.java]: nit for clarity: "When a location trait is ignored"

File: codegen/src/main/java/software/amazon/awssdk/codegen/AddShapes.java [hunk 12]
```diff
@@ -311,18 +311,67 @@ private ParameterHttpMapping generateParameterHttpMapping(Shape parentShape,
...
+        // https://smithy.io/2.0/spec/http-bindings.html
+        Location location = resolveLocation(parentShape, member, allC2jShapes);
```
Comment (alextwoods) [inline codegen/src/main/java/software/amazon/awssdk/codegen/AddShapes.java]: naming nit for clarity, maybe call this `resolvedLocation` (I think that makes the next line a bit easier to read.  It currently hurts my brain slightly)

File: codegen/src/main/java/software/amazon/awssdk/codegen/AddShapes.java (hunk 9 above)
Comment (alextwoods) [reply codegen/src/main/java/software/amazon/awssdk/codegen/AddShapes.java]: Or maybe just isTopLevelShape?

File: codegen/src/main/java/software/amazon/awssdk/codegen/AddShapes.java [hunk 13]
```diff
@@ -311,18 +311,67 @@ private ParameterHttpMapping generateParameterHttpMapping(Shape parentShape,
...
+                return isDirectOutputShape(parentShape, allC2jShapes) ? location : null;
+            default:
+                return location;
+        }
+    }
+
+    private boolean isDirectInputShape(Shape shape, Map<String, Shape> allC2jShapes) {
+        return builder.getService().getOperations().values().stream()
+                      .filter(o -> o.getInput() != null)
+                      .anyMatch(o -> allC2jShapes.get(o.getInput().getShape()).equals(shape));
```
Comment (alextwoods) [inline codegen/src/main/java/software/amazon/awssdk/codegen/AddShapes.java:358]: This probably isnt a big deal (but I ran into it in the Smithy POC...) - the `equals` method in shape is just a reference check.  That is probably actually what we want here so its fine.  But would be an issue if the shape itself was duplicated or something.

File: codegen/src/test/java/software/amazon/awssdk/codegen/AddShapesTest.java [hunk 14]
```diff
@@ -84,8 +84,29 @@ void generateShapeModel_memberRequiredByNestedShape_setsMemberModelAsRequired()
...
+    @Test
+    void generateShapeModel_locationOnDirectInputShape_isPreserved() {
+        ShapeModel inputShape = intermediateModel.getShapes().get("QueryParameterOperationRequest");
+        assertThat(inputShape.findMemberModelByC2jName("PathParam").getHttp().getLocation()).isEqualTo(Location.URI);
+        assertThat(inputShape.findMemberModelByC2jName("QueryParamOne").getHttp().getLocation()).isEqualTo(Location.QUERY_STRING);
+        assertThat(inputShape.findMemberModelByC2jName("StringHeaderMember").getHttp().getLocation()).isEqualTo(Location.HEADER);
+    }
+
+    @Test
+    void generateShapeModel_locationNameOnNestedShape_usesMemberNameForMarshalling() {
```
Comment (alextwoods) [inline codegen/src/test/java/software/amazon/awssdk/codegen/AddShapesTest.java:107]: Do we have the inverse of this in this test class (location null, locationName set)?

Comment (RanVaknin) [reply codegen/src/test/java/software/amazon/awssdk/codegen/AddShapesTest.java:107]: added the inverse test

File: codegen/src/main/java/software/amazon/awssdk/codegen/AddShapes.java (hunk 12 above)
Comment (RanVaknin) [reply codegen/src/main/java/software/amazon/awssdk/codegen/AddShapes.java]: done.

File: codegen/src/main/java/software/amazon/awssdk/codegen/AddShapes.java (hunk 10 above)
Comment (RanVaknin) [reply codegen/src/main/java/software/amazon/awssdk/codegen/AddShapes.java]: done.

File: codegen/src/main/java/software/amazon/awssdk/codegen/AddShapes.java [hunk 15]
```diff
@@ -342,38 +391,52 @@ private boolean isRequiredMember(String memberName, Shape memberShape) {
...
+        }
+
+        String requestUri = operation.get().getHttp().getRequestUri();
+        if (requestUri == null) {
+            String shapeName = allC2jShapes.entrySet().stream()
+                .filter(e -> e.getValue().equals(parentShape))
+                .map(Map.Entry::getKey)
+                .findFirst()
+                .orElseThrow(() -> new IllegalStateException("Shape not found in model: " + parentShape));
+            String detailMsg = "Operation referencing input shape '" + shapeName
```
Comment (joviegas) [inline codegen/src/main/java/software/amazon/awssdk/codegen/AddShapes.java:430]: Do we have any test cases asserting this error and error message so that shape name is printed correctly ?

Comment (RanVaknin) [reply codegen/src/main/java/software/amazon/awssdk/codegen/AddShapes.java:430]: Added 👍

**Review by joviegas — APPROVED** [review]
